File: C++ Programs/Algorithms/SortingAlgo/MergeSort.cpp
#include <iostream>
#include <vector>
using namespace std;

void disp(int *arr, int size)
{
    cout << "Sorted Array : ";
    for (int i = 0; i < size; i++)
    {
        if (i < (size - 1))
            cout << arr[i] << ", ";
        else
            cout << arr[i];
    }
}

void merge(int *arr, int low, int mid, int high)
{
    int size_left = mid - low + 1;
    int size_right = high - mid;
 
    int *left = new int[size_left],
        *right = new int[size_right];
 
    for (int i = 0; i < size_left; i++)
        left[i] = arr[low + i];
    for (int j = 0; j < size_right; j++)
        right[j] = arr[mid + 1 + j];
 
    int i = 0, j = 0;
    int k = low;
 
    while (i < size_left && j < size_right) {
        if (left[i] <= right[j]) {
            arr[k] = left[i];
            i++;
        }
        else {
            arr[k] = right[j];
            j++;
        }
        k++;
    }
    while (i < size_left) {
        arr[k] = left[i];
        i++;
        k++;
    }
    while (j < size_right) {
        arr[k] = right[j];
        j++;
        k++;
    }
    delete[] left;
    delete[] right;
}

void merge_sort(int arr[], int low, int high)
{
    if (low >= high)
        return;
    
    int mid = low + (high - low) / 2;
    merge_sort(arr, low, mid);
    merge_sort(arr, mid + 1, high);
    merge(arr, low, mid, high);
}

int main()
{   
    int size, data;
    cout << "Enter size of the list : ";
    cin >> size;
    int *arr = new int[size];
    cout << "Enter elements for the list (Unordered) : ";
    for (int i = 0; i < size; i++)
    {
        cin >> data;
        arr[i] = data;
    }
    merge_sort(arr, 0, size-1);
    disp(arr, size);
    return 0;
}

File: Java_Programs/Algorithms/SortingAlgo/MergeSort.java
package Java_Programs.Algorithms.SortingAlgo;

class MergeSort {
    void merge(int arr[], int low, int mid, int high) {
        int n1 = mid - low + 1;
        int n2 = high - mid;

        int L[] = new int[n1];
        int M[] = new int[n2];

        for (int i = 0; i < n1; i++)
            L[i] = arr[low + i];
        for (int j = 0; j < n2; j++)
            M[j] = arr[mid + 1 + j];

        int i, j, k;
        i = 0;
        j = 0;
        k = low;
        while (i < n1 && j < n2) {
            if (L[i] <= M[j]) {
                arr[k] = L[i];
                i++;
            } else {
                arr[k] = M[j];
                j++;
            }
            k++;
        }
        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
        }
        while (j < n2) {
            arr[k] = M[j];
            j++;
            k++;
        }
    }

    void mergeSort(int arr[], int low, int high) {
        if (low < high) {

            int m = (low + high) / 2;
            mergeSort(arr, low, m);
            mergeSort(arr, m + 1, high);
            merge(arr, low, m, high);
        }
    }

    static void printArray(int arr[]) {
        int n = arr.length;
        for (int i = 0; i < n; ++i)
            System.out.print(arr[i] + " ");
        System.out.println();
    }

    public static void main(String args[]) {
        int arr[] = { 6, 5, 12, 10, 9, 1 };
        MergeSort ob = new MergeSort();
        ob.mergeSort(arr, 0, arr.length - 1);
        System.out.println("Sorted array:");
        printArray(arr);
    }
}